atf_metrics/src/atf_metrics/calculate_user_result.py

```diff
@@ -88,82 +88,56 @@ def get_result(self):
-            
-            #print "groundtruth data is set from user within atf application for testblock %s. Skipping groundtruth evaluation from test_config"%self.testblock_name
-
-            # use data from user result
-            metric_result = self.metric_result
-
-            # overwrite user_result data with mandatory ATF filds
-            metric_result.name = self.name
-            metric_result.started = True
-            metric_result.finished = True
-            metric_result.groundtruth.available = True
-            #print "EXIT 1", metric_result.data.data, metric_result.groundtruth.result, metric_result.groundtruth.error_message
+        if not self.finished:
+            error_message = "testblock %s never stopped"%self.testblock_name
+            #print error_message
+            metric_result.groundtruth.result = False
+            metric_result.groundtruth.error_message = error_message
+            return metric_result
+
+        # check if stop was called
+        if self.metric_result == None:
+            error_message = "YOU SHOULD NOT END UP HERE: testblock %s stopped but metric_result is None"%self.testblock_name
+            #print error_message
+            metric_result.groundtruth.result = False
+            metric_result.groundtruth.error_message = error_message
             return metric_result
-        
-        #print "no groundtruth set via user_result", self.metric_result.groundtruth
-
-        metric_result.groundtruth.available = self.groundtruth.available
-        metric_result.groundtruth.data = self.groundtruth.data
-        metric_result.groundtruth.epsilon = self.groundtruth.epsilon
-        
-        # assign default value
-        metric_result.groundtruth.result = None
-        metric_result.groundtruth.error_message = None
-
-        if metric_result.started and metric_result.finished: #  we check if the testblock was ever started and stopped
-            # calculate metric data
-            # check if user has set any metric_result (not all default values anymore)
-            if not self.metric_result.started\
-                and not self.metric_result.finished\
-                and len(self.metric_result.series) == 0\
-                and self.metric_result.data.stamp == rospy.Time(0)\
-                and self.metric_result.data.data == 0:
-
-                # let the analyzer know that this test failed
-                metric_result.groundtruth.result = False
-                metric_result.groundtruth.error_message = "user result for testblock %s not set"%self.testblock_name
-                #print "EXIT 2", metric_result.data.data, metric_result.groundtruth.result, metric_result.groundtruth.error_message
-                return metric_result
 
-            metric_result.data = self.metric_result.data
+        # check if result is available
+        if metric_result.groundtruth.error_message.startswith("!!USER ERROR!!: no user result set"): # TODO use from global field (same as in atf.stop())
```

Comment:
I dont think we need this block because in this case, metric_result is already sufficiently filled (see my comment above) and will correctly be handled as any other case where result is set to false...right?

Reply:
I'm still in favor of keeping this block as it gives a different failure reason. I calling `atf.stop()` without user result we se the following failure reasons:

without this block
```
Failed ATF tests:
 - test 'ts0_c0_r0_e0_s0_0' (test1, robot1, env1, testblockset1): 
   - testblock 'testblock_5s': 
     - metric 'user_result::0': groundtruth missmatch: 0.000000 not within 0.800000+-0.200000
```

with this block
```
Failed ATF tests:
 - test 'ts0_c0_r0_e0_s0_0' (test1, robot1, env1, testblockset1): 
   - testblock 'testblock_5s': 
     - metric 'user_result::0': testblock testblock_5s stopped without user_result
```

Reply:
I'm asking myself now why `result` and `error_message` are part of the `Groundtruth.msg` and not directly under `MetricResult.msg`

with your explanations from yesterday, I now understand `groundtruth` is rather optional and contains only the "target" value and the epsilon range

having `result` and `error_message` in `Groundtruth.msg` doesn't seem to make too much sense

I rather think those two fields should be put (on level up) into `MetricResult` directly
as `result` and `error_message` are not necessarily groundtruth-related but are mandatory (at least `result`) also for metrics that do NOT use groundtruth

if you then would chose `uint8` (together with some enums) for `result` instead of `bool` - or even `actionlib_msgs.GoalStatus` directly, then the default value means something `PENDING`/`UNINITIALIZED`/`NOT_SET` and would make it easier to distinguish `result` not being set from `result` being negative/failure or positive/success

so any metric would not try to set the value of `metric_result.groundtruth.result` but rather `metric_result.result` - or whatever the field is called then


we can continue yesterday's discussion in another daily if you like